### PR TITLE
[#1512] Added new ember-qunit 4.3.0+ test helper

### PR DESCRIPTION
**Pull request by 0xadada**
If a consuming application is using `ember-cli-qunit` [4.2.0 or greater](https://github.com/ember-cli/ember-cli-qunit/blob/master/CHANGELOG.md#v420-2017-12-17), they may want to migrate to the
[more modern testing syntax](https://dockyard.com/blog/2018/01/11/modern-ember-testing). In that case, they will need to be imported from the `ember-simple-auth` addon namespace rather than from the consuming application namespace.

```js
// tests/acceptance/…
import { currentSession, authenticateSession, invalidateSession} from 'ember-simple-auth/test-support';
```

**Note** The helpers imported from the addon namespace (i.e. `import { functionName } from 'ember-simple-auth/test-support';`) differ slightly from those in the consumer application namespace in that they should not take the `application` argument.

The new-style helpers have the following function signatures:
* `currentSession()` returns the current session.
* `authenticateSession(sessionData)` authenticates the session; the optional `sessionData` argument can be used to mock an authenticator response (e.g. a token or user).
* `invalidateSession()` invalidates the session.

Based on suggestions by @cibernox added a new modern style helper, that
will be exported to the ember-simple-auth namespace

* [x] New test helper
* [x] Updated README.md to explain new helper
* [x] updated dependencies
* [x] Updated test-helper to use `setApplication`

Closes https://github.com/simplabs/ember-simple-auth/issues/1531
Fixes https://github.com/simplabs/ember-simple-auth/issues/1512

thanks @rwjblue @cibernox @marcoow @Turbo87 for helping me land this PR!

### CODE REVIEW

**Comment by rwjblue** [comment]
The newer style helpers do not work before Ember 2.4, I believe that you need to guard those tests (preventing them from running on 1.13 and 2.0).

For example [this](https://github.com/emberjs/ember-test-helpers/blob/7a92c721f8e55096a7bdb217e7cd294102aa82f8/tests/unit/setup-context-test.js#L25-L27) is what ember-test-helpers does to prevent its own test suite from failing on these older versions:

```js
import hasEmberVersion from 'ember-test-helpers/has-ember-version';

// ...snip...

describe('.....', function() {
  if (!hasEmberVersion(2, 4)) {
    return;
  }

  // ....rest of the tests here...
});
```

**Comment by 0xadada** [comment]
@rwjblue now guard against testing releases before 2.4 in 
https://github.com/simplabs/ember-simple-auth/pull/1536/commits/8dd43903b4a492fd7f3529103c49c83019b7a5d4

File: addon-test-support/index.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * @param {Object} sessionData Optional argument used to mock an authenticator
+ * response (e.g. a token or user).
+ * @return {Promise}
+ * @public
+ */
+export function authenticateSession(sessionData) {
+  const { owner } = getContext();
+  const session = owner.lookup(SESSION_SERVICE_KEY);
+  owner.setupRouter(); // router must initialize fully before authentication
```
Comment (rwjblue) [inline addon-test-support/index.js]: The prior versions of `authenticateSession` didn't do this, so I'm not 100% sure why this is needed.  Can you try to remove it and see what happens?

File: tests/acceptance/authentication-qunit-test.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  describe('the protected route', function() {
+    if (!hasEmberVersion(2, 4)) {
+      // guard against running test module on unsupported version (before 2.4)
+      return;
+    }
+
+    it('cannot be visited when the session is not authenticated', function() {
+      return invalidateSession().then(function() {
+        return visit('/protected').then(() => {
```
Comment (rwjblue) [inline tests/acceptance/authentication-qunit-test.js]: Can probably detangle these few tests a tad bit by returning early, something like:

```js
    it('cannot be visited when the session is not authenticated', function() {
      return invalidateSession()
        .then(() => visit('/protected'))
        .then(() => {
          expect(currentURL()).to.eq('/login');
        });
      });
    });
```

File: addon-test-support/index.js (hunk 1 above)
Comment (0xadada) [reply addon-test-support/index.js]: I added `setupRouter` because visits (`visit()`) to routes protected by `AuthenticatedRouteMixin` fail without it.

TL;DR `hasRoute` is undefined when the authenticationSucceeded event is triggered by the `ApplicationRouteMixin`. I did a [full writeup of the series of calls](https://github.com/simplabs/ember-simple-auth/pull/1519#issuecomment-359837396) describing what occurs on the previous PR. @jordpo found a workaround was to invoke `visit` first to setup the router, but by calling `setupRouter` in the test-helper, it would initialize the routerMicroLib and fix the problem.

**Comment by 0xadada** [comment]
@rwjblue looks nicer this way https://github.com/simplabs/ember-simple-auth/pull/1536/commits/872d97cd3f7b8d3b6a18c5f0607b2758b6a08cc0

File: addon-test-support/index.js (hunk 1 above)
Comment (rwjblue) [reply addon-test-support/index.js]: Thank you for pointing me at that writeup!  Ultimately, this still seems like a bug being identified. IMHO, it is not logical for the `ApplicationRouteMixin` to call `.transitionTo` before the first navigation (e.g. before the first `visit()` call)...

/cc @marcoow @Turbo87

Comment (marcoow) [reply addon-test-support/index.js]: > IMHO, it is not logical for the ApplicationRouteMixin to call .transitionTo before the first navigation (e.g. before the first visit() call)...

Yep, that's true. The problem is the [`ApplicationRouteMixin`'s `sessionAuthenticated` method](http://ember-simple-auth.com/api/classes/ApplicationRouteMixin.html#method_sessionAuthenticated) is called when the session is authenticated. Whether it makes sense to always do that automatically is certainly something that's debatable but that's how it is for now and we can't change that for now. We could maybe add a check into that method to just bail out early when the router is not ready (which I assume would require using private API to find out as well though 😞)

**Comment by 0xadada** [comment]
@marcoow The `setupRouter` function already does that, so I opted not to add the additional complexity.

-------- Original Message --------

File: addon-test-support/index.js (hunk 1 above)
Comment (0xadada) [reply addon-test-support/index.js]: @marcoow The `setupRouter` function already does that, so I opted not to add the additional complexity.

Comment (marcoow) [reply addon-test-support/index.js]: @0xadada: what I meant is we can add a check for whether the router is ready [here](https://github.com/simplabs/ember-simple-auth/blob/1.4.2/addon/mixins/application-route-mixin.js#L108) and just return if it is not.

File: tests/acceptance/authentication-qunit-test.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import { expect } from 'chai';
+import Pretender from 'pretender';
+import {
+  invalidateSession,
+  authenticateSession,
+  currentSession
+} from 'ember-simple-auth/test-support';
+import config from '../../config/environment';
+
+describe('Acceptance: Authentication (for QUnit helpers)', function() {
```
Comment (cibernox) [inline tests/acceptance/authentication-qunit-test.js]: Just a nit-pick, but those helpers are not QUnit specific (and as a matter of fact, in there they are working in mocha).
I'd name the module something different, like "New-style Test Helpers"

File: addon-test-support/index.js (hunk 1 above)
Comment (cibernox) [reply addon-test-support/index.js]: Is this the last loose end?

Comment (marcoow) [reply addon-test-support/index.js]: Yes, this is the last loose end. I guess we can leave it as is and see whether we can refactor the `ApplicationRouteMixin` separately so the error situation would never occur - that should be done in a separate PR anyways.

Comment (rwjblue) [reply addon-test-support/index.js]: Yes, totally agreed @marcoow. Calling `setupRouter` here should be fine for now. 

Let's open an issue to track refactoring the mixin so this isn't needed though...

File: tests/acceptance/authentication-qunit-test.js (hunk 3 above)
Comment (0xadada) [reply tests/acceptance/authentication-qunit-test.js]: @cibernox I'll rename the file and `describe('Acceptance: New-style Test Helpers)'.

**Comment by 0xadada** [comment]
Y'all, it feels so great to contribute back to the community. I feel like i've actually earned the ESA sticker i got at Ember Conf last year.

screen shot 2018-02-13 at 1 31 49 pm

Thanks to everyone who helped me land this PR, such a great community.
🎉

**Comment by marcoow** [comment]
Thanks so much for the effort @0xadada! This wasn't easy but I'm glad it's in now (once the build completes). Please say hi in Portland if you're there!

File: addon-test-support/index.js (hunk 1 above)
Comment (marcoow) [reply addon-test-support/index.js]: follow-up issue is here: https://github.com/simplabs/ember-simple-auth/issues/1539

**Comment by marcoow** [comment]
…and the build failed 😞 I restarted it but it doesn't look like a random failure…

**Comment by marcoow** [comment]
The problem is that `Ember.testing` is `false` [here](https://github.com/simplabs/ember-simple-auth/blob/master/addon/initializers/setup-session.js#L10).

**Comment by 0xadada** [comment]
@marcoow i'm confused about what changed between yesterday and today that would've suddenly started causing this?

**Comment by marcoow** [comment]
This must be some change in an upstream dependency. We ignore the `yarn.lock` file on CI so you'll get the latest of everything which sometimes can lead to errors.

**Comment by 0xadada** [comment]
@marcoow  the same error is failing on the `master` branch on my local system as well

**Comment by 0xadada** [comment]
looks like once [my tests `tearDownContext(context)`](https://github.com/simplabs/ember-simple-auth/pull/1536/files#diff-44b9f882bba69fbf5ee9455fdc5c781dR37) runs, it invokes [`@ember/test-helpers#tearDownContext`](https://github.com/emberjs/ember-test-helpers/blob/master/addon-test-support/%40ember/test-helpers/teardown-context.js#L30) which sets `Ember.testing = false`, and borks the other tests that use that.

File: tests/unit/initializers/setup-session-test.js [hunk 4]
```diff
@@ -19,6 +19,7 @@ describe('setupSession', () => {
       register() {},
       injection() {}
     };
+    Ember.testing = true;
```
Comment (0xadada) [inline tests/unit/initializers/setup-session-test.js]: Not really feelin' this, but `tearDownContext()` in the `new-style-helpers-test` blows this test up.

Comment (rwjblue) [reply tests/unit/initializers/setup-session-test.js]: Seems fine to me. I do dislike that Ember-test-helpers has to do that...

**Comment by marcoow** [comment]
So here we go - thanks for all the hard work @0xadada!

**Comment by marcoow** [comment]
This has been released with the 1.5.0 release: https://github.com/simplabs/ember-simple-auth/releases/tag/1.5.0
